aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -1,118 +1,154 @@
 
 using namespace vec256;
 
-// This adds the content of arr to sum
-template <class scalar_t, template <class> class OP, CPUCapability C>
-inline scalar_t allreduce_kernel_(const scalar_t *arr, size_t start, size_t end,
-                                  scalar_t sum) {
-  Vec256<scalar_t> part_sum;
-  // Use all 16 registers.
-  Vec256<scalar_t> tmp_sum[4], tmp_sum1, tmp_sum2, tmp_sum3;
-  Vec256<scalar_t> a[8];
-  size_t width =
-      256 / sizeof(scalar_t); // primitives per 256 bytes (two cache lines)
-  size_t epr = 32 / sizeof(scalar_t); // primitives per Vec256
-  size_t k = 0;
-  for (; k < (end - start) / width; k++) {
-    for (size_t i = 0; i < 8; i++) {
-      a[i].load(arr + (k * width) + i * epr + start);
+static inline int64_t round_down(int64_t a, int64_t m) {
+  return a - (a % m);
+}
+
+template<typename F>
+static void parallel_for(int64_t end, int64_t step, bool parallelize, F func) {
+  if (parallelize) {
+    tbb::parallel_for<int64_t>(0, end, step, func);
+  } else {
+    for (int64_t i = 0; i != end; i += step) {
+      func(i);
     }
-    for (size_t i = 0; i < 8; i += 2) {
-      tmp_sum[i / 2] = OP<Vec256<scalar_t>>()(a[i], a[i + 1]);
+  }
+}
+
+static tbb::affinity_partitioner ap;
+
+// Vectorized reduction defined by reduce operation `Op` with identity `ident`.
+// The reduction is built on top of reduce128, which reduces down a column
+// 128 bytes wide (WIDTH scalar elements). The width of 128 bytes is chosen
+// because of the "adjacent cache line prefetch" behavior on x86 CPUs.
+template<typename scalar_t, template <class> class Op, int ident>
+struct Reduction {
+  // reduction width in number of scalar elements
+  static constexpr int WIDTH = 128 / sizeof(scalar_t);
+
+  using Vec = Vec256<scalar_t>;
+  using Reduce = Op<Vec>;
+  using ReduceScalar = Op<scalar_t>;
+
+  static void apply(Tensor& res, const Tensor& self, at::optional<int64_t> dim) {
+    internal::init_tbb_num_threads();
+
+    auto out = res.data<scalar_t>();
+    auto data = self.data<scalar_t>();
+    auto numel = self.numel();
+    if (!dim.has_value()) {
+      *out = reduce_all(data, numel);
+      return;
     }
-    tmp_sum1 = OP<Vec256<scalar_t>>()(tmp_sum[0], tmp_sum[1]);
-    tmp_sum2 = OP<Vec256<scalar_t>>()(tmp_sum[2], tmp_sum[3]);
-    if (k == 0) {
-      part_sum = OP<Vec256<scalar_t>>()(tmp_sum1, tmp_sum2);
+
+    int64_t n = self.size(*dim);
+    int64_t stride = self.stride(*dim);
+    int64_t batch = numel / (n * stride);
+    bool paralellize = batch * n > internal::TBB_GRAIN_SIZE;
+    parallel_for(batch, 1, paralellize, [=](int64_t b) {
+      if (stride == 1) {
+        out[b] = reduce_all(&data[b * n], n);
+      } else {
+        reduce2d(&data[b * n * stride], &out[b * stride], n, stride, stride);
+      }
+    });
+  }
+
+  static scalar_t reduce_all(const scalar_t* data, int64_t size) {
+    int64_t k = size / WIDTH;
+
+    scalar_t sum;
+    if (size > internal::TBB_GRAIN_SIZE) {
+      sum = tbb::parallel_reduce(
```

Comment:
This need to call the tbb init function as otherwise the scheduler isn't setup correctly, which is why all calls to tbb were aggregated in parallel.h/.cpp

Reply:
init_tbb_num_threads is called by Reduction::apply which is the only entry point

Reply:
I'd add it simply as a convention, but I can't force you. The computational cost is negligible  and the cost of someone calling a tbb algorithm before the init justifies the extra line of code.